src/goto-instrument/contracts/utils.cpp

```diff
@@ -450,81 +450,114 @@ void add_quantified_variable(
   }
 }
 
-void replace_history_parameter(
+static void replace_history_parameter_rec(
   symbol_table_baset &symbol_table,
   exprt &expr,
-  std::map<exprt, exprt> &parameter2history,
-  source_locationt location,
+  std::unordered_map<exprt, symbol_exprt, irep_hash> &parameter2history,
+  const source_locationt &location,
   const irep_idt &mode,
   goto_programt &history,
-  const irep_idt &id)
+  const irep_idt &history_id)
 {
   for(auto &op : expr.operands())
   {
-    replace_history_parameter(
-      symbol_table, op, parameter2history, location, mode, history, id);
+    replace_history_parameter_rec(
+      symbol_table, op, parameter2history, location, mode, history, history_id);
   }
 
-  if(expr.id() == ID_old || expr.id() == ID_loop_entry)
-  {
-    const auto &parameter = to_history_expr(expr, id).expression();
+  if(expr.id() != ID_old && expr.id() != ID_loop_entry)
+    return;
 
-    const auto &id = parameter.id();
-    if(
-      id == ID_dereference || id == ID_member || id == ID_symbol ||
-      id == ID_ptrmember || id == ID_constant || id == ID_typecast ||
-      id == ID_index)
-    {
-      auto it = parameter2history.find(parameter);
+  const auto &parameter = to_history_expr(expr, history_id).expression();
+  const auto &id = parameter.id();
+  if(
+    id != ID_dereference && id != ID_member && id != ID_symbol &&
+    id != ID_ptrmember && id != ID_constant && id != ID_typecast &&
+    id != ID_index)
+  {
+    throw invalid_source_file_exceptiont(
+      "Tracking history of " + id2string(parameter.id()) +
+        " expressions is not supported yet.",
+      expr.source_location());
+  }
```

Comment:
You could add the following regression test to cover this case.
```c
void foo(int *x)
  __CPROVER_assigns(x)
  __CPROVER_ensures(__CPROVER_POINTER_OFFSET(x) == __CPROVER_old(__CPROVER_POINTER_OFFSET(x)) + 1)
  __CPROVER_ensures(*x == __CPROVER_old(*x) + 1)
{
  *x = *x + 1;
  x++;
}

int main()
{
  int n[2];
  foo(&n);

  return 0;
}
```

Reply:
Thank you for this suggestion, but the history-invalid regression test actually seeks to cover this. It will, however, not trigger this piece of code as the C front-end already picks up the problem under the very same condition. I will, therefore, change this exception to an `INVARIANT`.